src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,196 @@
+    // update expr loop id
+    remove_last_loop_id(expr);
+    // move if it is not the first
+    if (expr != *inner_loop_begin_pos) {
+        auto port_expr_iter = std::find(inner_loop_begin_pos, inner_loop_end_pos, expr);
+        OPENVINO_ASSERT(port_expr_iter != inner_loop_end_pos, "Identified extractable expr is not found in loop.");
+        linear_ir.move(port_expr_iter, inner_loop_begin_pos);
+    } else {
+        inner_loop_begin_pos++;
+    }
+}
+
+void update_loop_ports(const ExpressionPtr& expr, const LoopManagerPtr& loop_manager, size_t inner_loop_id,
+                       const LinearIR::constExprIt& inner_loop_begin_pos, const LinearIR::constExprIt& inner_loop_end_pos) {
+    const auto& inner_loop_info = loop_manager->get_loop_info<UnifiedLoopInfo>(inner_loop_id);
+    // delete expr input ports from loop input points, add expr output ports' consumers if
+    // consumed in inner loop to loop input ports.
+    std::vector<ExpressionPort> new_loop_input_ports;
+    for (size_t i = 0; i < expr->get_output_count(); i++) {
+        const auto& consumers = expr->get_output_port_connector(i)->get_consumers();
+        for (const auto& consumer : consumers) {
+            const auto& loop_ids = consumer.get_expr()->get_loop_ids();
+            if (std::find(loop_ids.cbegin(), loop_ids.cend(), inner_loop_id) != loop_ids.cend()) {
+                new_loop_input_ports.push_back(consumer);
+            }
+        }
+    }
+    const auto& expr_input_ports = expr->get_input_ports();
+    inner_loop_info->update_loop_ports(expr_input_ports, new_loop_input_ports);
+
+    // delete expr out ports from loop out ports directly if it's in loop output ports
+    std::vector<ExpressionPort> out_ports_to_delete;
+    for (size_t i = 0; i < expr->get_output_count(); ++i) {
+        const auto& out_port = expr->get_output_port(i);
+        if (inner_loop_info->is_loop_port(out_port)) {
+            out_ports_to_delete.push_back(out_port);
+        }
+    }
+    if (!out_ports_to_delete.empty()) {
+        std::vector<ExpressionPort> new_ports;
+        inner_loop_info->update_loop_ports(out_ports_to_delete, new_ports);
+    }
+    // TODO: 142990.
+    // Need sort after update loop ports. There are possibility that all exprs are moved to outer loop.
+    if (!inner_loop_info->get_input_ports().empty() && !inner_loop_info->get_output_ports().empty()) {
+        loop_manager->sort_loop_ports(inner_loop_begin_pos, inner_loop_end_pos, inner_loop_id);
+    }
+}
+
+std::vector<ExpressionPtr> get_loop_input_exprs(const std::vector<LoopPort>& loop_in_ports) {
+    std::vector<ExpressionPtr> input_exprs;
+    for (size_t port_num = 0; port_num < loop_in_ports.size(); ++port_num) {
+        const auto& expr = loop_in_ports[port_num].expr_port->get_expr();
+        if (std::find(input_exprs.begin(), input_exprs.end(), expr) == input_exprs.end()) {
+            input_exprs.push_back(expr);
+        }
```

Comment:
This algorithm has O(N^2) complexity with respect to the `loop_in_ports` size.
It can be easily improved to O(N) if we keep both unordered_set to check if we saw this expression before and vector to store the seen expressions in the original order:
```suggestion
        if (seen_exprs.count(expr) == 0) {
            seen_exprs.insert(expr);
            input_exprs.push_back(expr);
        }
```

Reply:
updated, thanks Ivan!